src/Development/Shake/Internal/Core/Rules.hs

```diff
@@ -133,12 +134,12 @@ modifyRulesScoped f (Rules r) = Rules $ do
         modifyIORef' refOld (<> f rules)
         return res
 
-runRules :: ShakeOptions -> Rules () -> IO ([(Stack, Action ())], Map.HashMap TypeRep BuiltinRule, TMap.Map UserRuleVersioned, [Target])
+runRules :: ShakeOptions -> Rules () -> IO ([(Stack, Action ())], Map.HashMap TypeRep BuiltinRule, TMap.Map UserRuleVersioned, [Target], String)
```

Comment:
Now there are 4 fields it probably wants to be a real return data type, e.g. `data RulesInfo = {riActions :: [(Stack, Action ()]], ...}`. However, this isn't essential.